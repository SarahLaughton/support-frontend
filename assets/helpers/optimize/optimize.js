// @flow

// ----- Imports ----- //

import * as storage from 'helpers/storage';

// ----- Types ----- //

export type OptimizeExperiment = {
  id: string,
  variant: string,
}

export type OptimizeExperiments = OptimizeExperiment[];

const OPTIMIZE_STORAGE_KEY = 'optimizeExperiments';

const EXPERIMENTS_UPDATED = 'OPTIMIZE_EXPERIMENTS_UPDATED';

// ----- Functions ----- //

function optimizeIsLoaded() {
  return window.dataLayer != null;
}

function gtag() {
  if (optimizeIsLoaded()) {
    // eslint-disable-next-line prefer-rest-params
    window.dataLayer.push(arguments); // unfortunately Optimize seems to need the Arguments object, not just an array
  }
}

function getExperimentsFromApi(callback: (variant: string, id: string) => void) {
  // $FlowIgnore
  gtag('event', 'optimize.callback', { callback });
}

function readExperimentsFromSession(): OptimizeExperiments {
  try {
    const data = storage.getSession(OPTIMIZE_STORAGE_KEY);
    return data ? JSON.parse(data) : [];
  } catch (e) {
    return [];
  }
}

function storeExperimentsInSession(experiments: OptimizeExperiments): boolean {
  try {
    const serialised = JSON.stringify(experiments);
    storage.setSession(OPTIMIZE_STORAGE_KEY, serialised);
    return true;
  } catch (_) {
    return false;
  }
}

function storeExperimentInSession(experiment: OptimizeExperiment): boolean {
  try {
    const existing: OptimizeExperiments = readExperimentsFromSession();
    const updated: OptimizeExperiments = existing
      .filter(exp => exp.id !== experiment.id)
      .concat(experiment);

    storeExperimentsInSession(updated);
    return true;
  } catch (_) {
    return false;
  }
}

function addExperimentUpdateListener(addToStoreCallback: OptimizeExperiment => void) {
  window.addEventListener(
    EXPERIMENTS_UPDATED,
    ev => addToStoreCallback({ id: ev.detail.id, variant: ev.detail.variant }),
  );
}

function addOptimizeExperiments(addToStoreCallback: OptimizeExperiment => void) {
  // Store experiments in the session as well as Redux
  const withSessionStorageCallback = (exp: OptimizeExperiment) => {
    storeExperimentInSession(exp);
    addToStoreCallback(exp);
  };

  if (optimizeIsLoaded()) {
    getExperimentsFromApi((variant, id) => withSessionStorageCallback({ id, variant }));
  } else {
    // Add a listener so we can update the store once Optimize loads
    addExperimentUpdateListener(withSessionStorageCallback);
  }
}

function getOptimizeExperiments() {
  getExperimentsFromApi((value, name) => {
    window.dispatchEvent(new CustomEvent(EXPERIMENTS_UPDATED, {
      detail: {
        id: name,
        variant: value,
      },
    }));
  });
}

// ----- Exports ----- //

export {
  addOptimizeExperiments,
  readExperimentsFromSession,
  getOptimizeExperiments,
};
