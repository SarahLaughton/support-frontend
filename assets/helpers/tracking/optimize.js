// @flow

// ----- Imports ----- //

import * as storage from 'helpers/storage';
import { getQueryParameter } from 'helpers/url';
import { deserialiseJsonObject } from 'helpers/utilities';
import { gaPropertyId } from './googleTagManager';


// ----- Types ----- //

type Experiment = string;
type Variant = string;

export type OptimizeExperiments = {
  [Experiment]: Variant,
};

const OPTIMIZE_STORAGE_KEY = 'optimizeExperiments';

const OPTIMIZE_QUERY_PARAMETER = 'utm_expid';


// ----- Functions ----- //

// Makes sure experiments are of the type [string]: string to match participations.
function parseExperimentsFromGaData(optimize: Object): OptimizeExperiments {

  return Object.keys(optimize).reduce((result, key) => {
    if (typeof optimize[key] === 'string') {
      return { ...result, [key]: optimize[key] };
    }

    return result;
  }, {});
}

function getExperimentsFromGaData(): OptimizeExperiments {
  try {
    return parseExperimentsFromGaData(window.gaData[gaPropertyId].experiments);
  } catch (_) {
    return {};
  }
}

// Returns an Optimize experiment from the query parameter utm_expid.
// Expected format of query parameter is '.<testid>.<variantid>'
function parseExperimentFromQueryParam(queryParameter: ?string): OptimizeExperiments {
  if (queryParameter) {
    // eslint-disable-next-line no-unused-vars
    const [_, experimentId, variantId] = queryParameter.split('.');
    if (experimentId && variantId) {
      return { [experimentId]: variantId };
    }
  }
  return {};
}

function storeOptimizeExperimentsInSession(experiments: OptimizeExperiments): boolean {
  try {
    const serialised = JSON.stringify(experiments);
    storage.setSession(OPTIMIZE_STORAGE_KEY, serialised);
    return true;
  } catch (_) {
    return false;
  }
}

function readOptimizeExperimentsInSession(): OptimizeExperiments {
  const data = storage.getSession(OPTIMIZE_STORAGE_KEY);
  return data ? deserialiseJsonObject(data) || {} : {};
}

function getOptimizeExperiments(): OptimizeExperiments {
  const experiments = {
    ...parseExperimentFromQueryParam(getQueryParameter(OPTIMIZE_QUERY_PARAMETER)),
    ...getExperimentsFromGaData(),
    ...readOptimizeExperimentsInSession(),
  };
  storeOptimizeExperimentsInSession(experiments);
  return experiments;
}


// ----- Exports ----- //

export {
  OPTIMIZE_QUERY_PARAMETER,
  getOptimizeExperiments,
  parseExperimentsFromGaData,
  parseExperimentFromQueryParam,
};
