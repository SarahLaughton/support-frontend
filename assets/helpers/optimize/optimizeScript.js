/* eslint-disable */
import { getOptimizeExperiments } from './optimize';

try {
  var db = indexedDB.open("test");
  // Check if Firefox Private Browsing is enabled
  // because the page hiding snippet doesn't work
  // properly in FF PB mode, see here:
  // https://www.en.advertisercommunity.com/t5/Google-Optimize-Implement/Optimize-Page-Hiding-Snippet-Unhide-delay-issue-in-Firefox/td-p/1106919
  db.onsuccess = function() {
    // Not in FF PB mode
    (function(a,s,y,n,c,h,i,d,e){s.className+=' '+y;h.start=1*new Date;
      h.end=i=function(){s.className=s.className.replace(RegExp(' ?'+y),'')};
      (a[n]=a[n]||[]).hide=h;setTimeout(function(){i();h.end=null},c);h.timeout=c;
    })(window,document.documentElement,'async-hide','dataLayer',4000,
      {'GTM-KGKKPS4':true, 'GTM-NZGXNBL':true});

    (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
    })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

    ga('create', 'UA-51507017-5', 'auto', {cookieDomain: 'auto', anonymizeIp: true});
    ga('require', 'GTM-NZGXNBL');
    ga('require', 'GTM-KGKKPS4');

    getOptimizeExperiments();
  };
} catch (e) {
  console.log(`Error initialising Optimize script: ${e.message}`);
}
/* eslint-enable */
