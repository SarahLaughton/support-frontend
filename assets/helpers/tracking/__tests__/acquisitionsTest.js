// @flow

// ----- Imports ----- //

import { type OptimizeExperiments } from '../optimize';

import {
  derivePaymentApiAcquisitionData,
  getOphanIds,
  participationsToAcquisitionABTest,
  optimizeToAcquisitionABTest,
  acquisitionsWithOptimize,
  type ReferrerAcquisitionData,
} from '../acquisitions';


// ----- Tests ----- //

jest.mock('ophan', () => ({ viewId: '123456' }));

describe('acquisitions', () => {

  describe('derivePaymentApiAcquisitionData', () => {

    it('should return a PaymentAPIAcquisitionData correctly built', () => {
      const referrerAcquisitionData = {
        campaignCode: 'Example',
        referrerPageviewId: '123456',
        referrerUrl: 'https://example.com',
        componentId: 'exampleComponentId',
        componentType: 'exampleComponentType',
        source: 'exampleSource',
        abTest: {
          name: 'referrerAbTest',
          variant: 'value1',
        },
        abTests: [],
        queryParameters: [{ name: 'param1', value: 'value1' }, { name: 'param2', value: 'value2' }],
      };

      const nativeAbParticipations = {
        testId1: 'variant1',
        testId2: 'variant2',
        testId3: 'variant3',
      };

      const paymentApiAcquisitionData =
        derivePaymentApiAcquisitionData(referrerAcquisitionData, nativeAbParticipations);

      expect(paymentApiAcquisitionData).toMatchSnapshot();

      // The abTests array should be a combination of supportAbTests and the source tests
      expect(paymentApiAcquisitionData.abTests && paymentApiAcquisitionData.abTests.length).toEqual(4);
      expect(paymentApiAcquisitionData.campaignCodes && paymentApiAcquisitionData.campaignCodes.length).toEqual(1);
    });
  });

  describe('getOphanIds', () => {

    it('should return null for the browserId and visitId when they are not present in the cookies', () => {

      const { pageviewId, browserId, visitId } = getOphanIds();

      expect(pageviewId).toBe('123456');
      expect(browserId).toBeNull();
      expect(visitId).toBeNull();
    });

    it('should read the browserId and visitId from cookie and pageViewId from ophan', () => {

      document.cookie = 'bwid=123';
      document.cookie = 'vsid=456';

      const { pageviewId, browserId, visitId } = getOphanIds();

      expect(pageviewId).toBe('123456');
      expect(browserId).toBe('123');
      expect(visitId).toBe('456');
    });
  });

  describe('participationsToAcquisitionABTest', () => {

    it('should return an empty array in the presence of a empty object as its input', () => {
      const participations = {};
      const acquisitionABTest = participationsToAcquisitionABTest(participations);
      expect(acquisitionABTest).toEqual([]);
    });

    it('should return an array of AcquisitionAbTests', () => {
      const participations = {
        test0: 'variant0',
        test1: 'variant1',
        test2: 'variant2',
      };
      const acquisitionABTests = participationsToAcquisitionABTest(participations);

      expect(acquisitionABTests.length).toEqual(3);
      expect(acquisitionABTests[0]).toEqual({ name: 'test0', variant: 'variant0' });
    });
  });

  describe('optimizeToAcquisitionABTest', () => {

    it('should return an empty array in the presence of a empty object as its input', () => {
      expect(optimizeToAcquisitionABTest({})).toEqual([]);
    });

    it('should return an array of AcquisitionAbTests appended with an Optimize tag', () => {

      const optimizeExperiments: OptimizeExperiments = {
        testOne: 'variantOne',
        testTwo: 'variantTwo',
        testThree: 'variantSeven',
      };

      const acquisitionABTests = optimizeToAcquisitionABTest(optimizeExperiments);

      expect(acquisitionABTests.length).toBe(3);
      expect(acquisitionABTests[0]).toEqual({ name: 'testOne$Optimize', variant: 'variantOne' });

    });

  });

  describe('acquisitionsWithOptimize', () => {

    it('should produce referrerAcquisitionData with optimize experiments', () => {

      const acquisitionData: ReferrerAcquisitionData = {
        campaignCode: 'Example',
        referrerPageviewId: '123456',
        referrerUrl: 'https://example.com',
        componentId: 'exampleComponentId',
        componentType: 'exampleComponentType',
        source: 'exampleSource',
        abTest: null,
        abTests: [{
          name: 'abTestOne',
          variant: 'valueOne',
        }],
        queryParameters: [{ name: 'param1', value: 'value1' }, { name: 'param2', value: 'value2' }],
      };

      const experiments: OptimizeExperiments = {
        testOne: 'variantOne',
        testTwo: 'variantTwo',
        testThree: 'variantSeven',
      };

      const updatedAcquisitionData = acquisitionsWithOptimize(acquisitionData, experiments);

      expect(updatedAcquisitionData).toMatchSnapshot();
      expect(updatedAcquisitionData.abTests && updatedAcquisitionData.abTests.length).toBe(4);
      expect(updatedAcquisitionData.abTests && updatedAcquisitionData.abTests[0]).toEqual({
        name: 'testOne$Optimize',
        variant: 'variantOne',
      });
      expect(updatedAcquisitionData.abTests && updatedAcquisitionData.abTests[3]).toEqual({
        name: 'abTestOne',
        variant: 'valueOne',
      });

    });

  });

});
