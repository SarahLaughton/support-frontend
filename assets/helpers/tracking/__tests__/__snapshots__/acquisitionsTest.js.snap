// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`acquisitions acquisitionsWithOptimize should produce referrerAcquisitionData with optimize experiments 1`] = `
Object {
  "abTest": null,
  "abTests": Array [
    Object {
      "name": "testOne$Optimize",
      "variant": "variantOne",
    },
    Object {
      "name": "testTwo$Optimize",
      "variant": "variantTwo",
    },
    Object {
      "name": "testThree$Optimize",
      "variant": "variantSeven",
    },
    Object {
      "name": "abTestOne",
      "variant": "valueOne",
    },
  ],
  "campaignCode": "Example",
  "componentId": "exampleComponentId",
  "componentType": "exampleComponentType",
  "queryParameters": Array [
    Object {
      "name": "param1",
      "value": "value1",
    },
    Object {
      "name": "param2",
      "value": "value2",
    },
  ],
  "referrerPageviewId": "123456",
  "referrerUrl": "https://example.com",
  "source": "exampleSource",
}
`;

exports[`acquisitions derivePaymentApiAcquisitionData should return a PaymentAPIAcquisitionData correctly built 1`] = `
Object {
  "abTests": Array [
    Object {
      "name": "testId1",
      "variant": "variant1",
    },
    Object {
      "name": "testId2",
      "variant": "variant2",
    },
    Object {
      "name": "testId3",
      "variant": "variant3",
    },
    Object {
      "name": "referrerAbTest",
      "variant": "value1",
    },
  ],
  "browserId": null,
  "campaignCodes": Array [
    "Example",
  ],
  "componentId": "exampleComponentId",
  "componentType": "exampleComponentType",
  "pageviewId": "123456",
  "platform": "SUPPORT",
  "referrerPageviewId": "123456",
  "referrerUrl": "https://example.com",
  "source": "exampleSource",
  "visitId": null,
}
`;
